FF1Lib/Magic.cs

```diff
@@ -456,6 +470,66 @@ public void ShuffleMagicLevels(MT19337 rng, bool keepPermissions, bool tieredShu
 			Put(ConfusedSpellIndexOffset, new[] { (byte)newFireSpellIndex });
 		}
 
+		public void SetMPMax(int redMageMaxMP, int whiteMageMaxMP, int blackMageMaxMP, int knightNinjaMaxMp)
+		{
+			//class index order: fighter, thief, blackbelt, redmage, whitemage, blackmage
+			SetClassMaxMp(3, redMageMaxMP);
+			SetClassMaxMp(4, whiteMageMaxMP);
+			SetClassMaxMp(5, blackMageMaxMP);
+
+			//for now, before we change the asm for knight and ninja, the two will be tied together
+			Put(KnightNinjaMaxMPOffset, new byte[] { (byte)(knightNinjaMaxMp + 1)});
+		}
+
+		public void SetClassMaxMp(int classIndex, int maxMp)
+		{
+			//49 levels per class, 2 bytes
+			//spell data is packed into a byte, bit index = spell level gained that level
+			//0 0 0 0 1 0 0 0 would mean gaining a level 4 spell slot that level
+
+			//brute force way... count the number of spells and rewrite it to not gain any more after the max
+			List<int> spellCount = new List<int> { 2, 0, 0, 0, 0, 0, 0, 0 };
+			for (int i = 0; i < 49; i++)
+			{
+				int currentOffset = NewLevelUpDataOffset + (49 * classIndex * 2) + (i * 2) + 1;
+				byte currentSpellData = Get(currentOffset, 1)[0];
+
+				for (int bitTest = 0; bitTest < 8; bitTest++)
+				{
+					if ((currentSpellData & (1 << bitTest)) != 0)
+					{
+						spellCount[bitTest]++;
+					}
+
+					if (spellCount[bitTest] > maxMp)
+					{
+						currentSpellData = (byte)(currentSpellData & ~(1 << bitTest));
+					}
+				}
+
+				Put(currentOffset, new byte[] { currentSpellData });
+			}
+		}
+
+		public void ChangeLockMode(LockHitMode lockHitMode)
+		{
+			//must be done before spells get shuffled around otherwise we'd be changing a spell that isnt lock

```

Comment:
this comment is more useful in the place this is called in FF1Rom.cs

Reply:
That does make a lot more sense, it could save somebody a lot of grief in the future. I've made that change.